### PR TITLE
JWT-auth: Support aud as string array

### PR DESCRIPTION
**Pull request by qiwzhang**
**What this PR does / why we need it**:

To support "aud" in JWT as string array.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by qiwzhang** [comment]
https://github.com/istio/proxy/issues/985

File: src/envoy/auth/jwt.cc [hunk 1]
```diff
@@ -277,10 +277,24 @@ Jwt::Jwt(const std::string &jwt) {
...
   iss_ = payload_->getString("iss", "");
-  aud_ = payload_->getString("aud", "");
   sub_ = payload_->getString("sub", "");
   exp_ = payload_->getInteger("exp", 0);
 
+  // "aud" can be either string array or string.
+  // Try as string array, read it as empty array if doesn't exist.
+  try {
+    aud_ = payload_->getStringArray("aud", true);
+  } catch (...) {
```
Comment (lizan) [inline src/envoy/auth/jwt.cc]: Do not catch all exceptions but `Json::Exception` only.

**Comment by lizan** [comment]
Can you also add `Fixes #985` to PR description so the issue will be auto closed?

File: src/envoy/auth/jwt.cc (hunk 1 above)
Comment (qiwzhang) [reply src/envoy/auth/jwt.cc]: Done

**Comment by sreetummidi** [comment]
@lizan & @qiwzhang I am still seeing the exception below in the istio proxy log with istio 0.5.1 release

terminate called after throwing an instance of 'Envoy::Json::Exception'
  what():  key 'aud' missing or not a string from lines 1-1
2018-02-14T23:30:00.368852Z	warn	Epoch 1 terminated with an error: signal: aborted (core dumped)

The JWT I am passing contains the claim
"aud": [
    "apps_manager_js"
  ]


My Policy contains the audience as 

    audiences:
    - apps_manager_js

**Comment by qiwzhang** [comment]
Sorry,  0.5.1 was cut 14 days ago.  My fix was in 9 days ago.
https://github.com/istio/proxy/pull/994

**Comment by sreetummidi** [comment]
Thank you :)

**Pull request merged**